### PR TITLE
QL: Eliminate internal type DATETIME_NANOS

### PR DESCRIPTION
**Pull request by matriv**
Moving towards grouping of data types in the field caps API
the internal data type `DATETIME_NANOS`, introduced for `date_nanos`
support, is eliminated.

Relates: #67722
Follows: #67666

### CODE REVIEW

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/extractor/FieldHitExtractor.java [hunk 1]
```diff
@@ -49,12 +49,12 @@ protected Object unwrapCustomValue(Object values) {
...
         if (dataType == DATETIME) {
             if (values instanceof String) {
-                return parseDateString(values);
-            }
-        }
-        if (dataType == DATETIME_NANOS) {
-            if (values instanceof String) {
-                return DateUtils.asDateTimeWithNanos(values.toString(), zoneId());
+                try {
+                    return DateUtils.asDateTimeWithNanos(values.toString(), zoneId());
```
Comment (costin) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/extractor/FieldHitExtractor.java:53]: I'd wait for the fields API from @astefan to land before going for this one.
Hopefully that eliminates the internal handling of data format and thus this try/catch as well.

**Review by costin — COMMENTED** [review]
Shouldn't the DATE_NANOS type be removed as well, per the PR description?
The main feedback that I have is for this PR to land after adding the transition to fields API.

**Comment by matriv** [comment]
> Shouldn't the DATE_NANOS type be removed as well, per the PR description?
> The main feedback that I have is for this PR to land after adding the transition to fields API.

No problem to wait until the fields API transition?

What do you mean `DATE_NANOS`? There's no such data type introduced in QL/SQL.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/extractor/FieldHitExtractor.java (hunk 1 above)
Comment (bpintea) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/extractor/FieldHitExtractor.java:53]: I guess fields API will make this uniform, but just in case, an alternative might be to check the presence of the `T` or `.` in the string and based on that call the right parser.

**Comment by mark-vieira** [comment]
@elasticmachine update branch

**Comment by costin** [comment]
> What do you mean DATE_NANOS? There's no such data type introduced in QL/SQL.

I meant `org.elasticsearch.xpack.ql.type.DataTypes.DATETIME_NANOS`

**Comment by matriv** [comment]
> > What do you mean DATE_NANOS? There's no such data type introduced in QL/SQL.
> 
> I meant `org.elasticsearch.xpack.ql.type.DataTypes.DATETIME_NANOS`

https://github.com/elastic/elasticsearch/pull/68220/files#diff-f31cd8869d3c81accd160cec419b0b8ae9308c0dcb4b5c9d38afedc0de6321a6L43

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/extractor/FieldHitExtractor.java (hunk 1 above)
Comment (matriv) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/extractor/FieldHitExtractor.java:54]: I was curious to see the perf difference of try/catch and trying to check the string:
```
Benchmark                              Mode  Cnt        Score      Error  Units
BenchmarkDateParse.testWithContains   thrpt   25  1631774,366 ± 6016,080  ops/s
BenchmarkDateParse.testWithException  thrpt   25   629767,727 ± 3843,439  ops/s
```
and it's quite big!

Comment (matriv) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/extractor/FieldHitExtractor.java:54]: I will use the `contains("-")` since it's always there for Datetime stings and it's at the beginning of the string.

Comment (bpintea) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/extractor/FieldHitExtractor.java]: > and it's quite big!

I wonder if indexing the string wouldn't actually make a further difference: contains takes each character at a time, but the datetime format is fixed, so `-` should always be at `.charAt(4)`. Either way, I think it's fine.

Comment (matriv) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/extractor/FieldHitExtractor.java]: hm, yeah, I rushed here, need to revisit:

- a timestamp can start with `-` (negative millis).
- `.charAt(4)` is also not safe, as there are possible dates like: `+14953-02-....` or `-134-05-....`

**Comment by matriv** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by matriv** [comment]
We don't need to extract date_nanos for `EQL` as we don't have a project phase as in `SQL`.
Could be needed if we implement support for `date_nanos` for sequence matching: Opened #68812 to track this.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/querydsl/container/SearchHitFieldRef.java [hunk 2]
```diff
@@ -64,10 +64,13 @@ public String toString() {
         return name;
     }
 
+    private static boolean hasDocValues(DataType dataType) {
+        return dataType == KEYWORD || dataType == DATETIME;
+    }
+
```
Comment (astefan) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/querydsl/container/SearchHitFieldRef.java:70]: This is not needed.

**Review by astefan — COMMENTED** [review]
LGTM in general. I've left one comment about a method that seems to have sneaked in at merge time.
Also, in `ExpressionTests` do we still need the `testCastToDatetimeNanosNotSupported()` test?

**Comment by matriv** [comment]
@astefan Nice catch, removing the test, and indeed I missed this merge mistake with the `hasDocValues()` method, removing.

**Review by bpintea — APPROVED** [review]
LGTM

**Review by astefan — APPROVED** [review]
LGTM

**Comment by astefan** [comment]
@elasticmachine run elasticsearch-ci/2

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/extractor/FieldHitExtractor.java [hunk 3]
```diff
@@ -50,20 +48,17 @@ protected Object unwrapCustomValue(Object values) {
...
             if (values instanceof String) {
-                try {
-                    return DateUtils.asDateTimeWithNanos(values.toString(), zoneId());
-                } catch (IllegalArgumentException | DateTimeParseException e) {
-                    // For bwc compatibility during rolling upgrade
-                    return parseDateString(values);
-                }
+                // We ask @timestamp (or the defined alternative field) to be returned as `epoch_millis`
+                // when matching sequence to avoid parsing into ZonedDateTime objects for performance reasons.
+                return parseEpochMillisAsString(values.toString());
```
Comment (costin) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/extractor/FieldHitExtractor.java:53]: +1 to this improvement.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java [hunk 4]
```diff
@@ -118,49 +120,7 @@ protected Object unwrapFieldsMultiValue(Object values) {
...
-
-    protected boolean isFromDocValuesOnly(DataType dataType) {
-        return dataType == KEYWORD // because of ignore_above.
-                    || dataType == DATETIME
-                    || dataType == SCALED_FLOAT; // because of scaling_factor
-    }
-    
-    private static NumberType numberType(DataType dataType) {
-        return NumberType.valueOf(dataType.esType().toUpperCase(Locale.ROOT));
+        return values;
```
Comment (costin) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java]: Isn't the left side removed due to the use of fields API? Just double checking why the change appears to belong to this PR in Github.

**Review by costin — APPROVED** [review]
More lines removed than added - that's nice.
Left a comment regarding the history in Github otherwise LGTM.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java (hunk 4 above)
Comment (matriv) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java]: It was a merge mistake, it's reverted already.
